### DISCUSSION TITLE
Is there any way to delete image from storage using sql?

### DISCUSSION THREAD

**Discussion by growthondev** [supabase/supabase#7067] [Questions] [answered]
AFAIK remove a row from storage.objects is just deleting a meta data. Is that correct?

How can I remove a actual image from storage using sql?

Thank you.

**Comment by GaryAustin1** [comment]
I use this function with the http extension when a note containing the image is deleted.  I have a trigger function that actually calls this one, but this could be in a trigger function directly.

```
create function delete_file(bucket character varying, pathname character varying) returns character varying
    security definer
    language plpgsql
as
$$
declare
  myurl varchar := 'https://zkinvaaaaaaqhblxrq.supabase.co';
  service_role_key varchar := 'eyxxxxxxxxxxxxxxxxxxIkpXVCJ9.eyJyb2xlIjoic2VydmljZV9yb2xlIiwiaWxxxxxxxxCJleHAiOjE5NDc0MDY0Njd9.EQKfltJHqjVsZBwsaYEKQYTZOnk_p5S8-xxxxxxxxx';
  url varchar := myurl||'/storage/v1/object/'||bucket||'/'||pathname;
begin
return (
select status FROM http((
          'DELETE',
            url,
           ARRAY[http_header('authorization','Bearer '||service_role_key)],
           NULL,
           NULL
        )::http_request)
);
end;
$$;
```
Note that you do have to put your credentials hardcoded.
EDIT:
You will need to add
SET search_path = public, extensions 
Unless you set up extensions in your API settings.

**Reply by markbulingit** [reply]
@GaryAustin1 this thing only works via API, when you delete a record via SQL the object on storage will still be there

**Reply by GaryAustin1** [reply]
Yes that is the reason the call is going to the API in my code above.  You can’t just delete the storage.objects row.  Here is a more advance solution I’ve been working on https://github.com/GaryAustin1/supa-file-helper

**Comment by milesingrams** [comment]
I made a feature request for these API improvements here:
https://github.com/supabase/supabase/issues/7939

Definitely feel like this is not just a nice-to-have but a must-have given how often storage is tied to database entries and needs to be able to be deleted using triggers.

**Comment by fenos** [comment, marked as answer]
Thanks for the feature request,

Currently is not something we'd like to support.
We believe that it's better to have a single source of truth for handling file operations (the API the layer)

Providing unidirectional operations from SQL seems to be something hard to maintain.

I'm open for more idea's, let's continue on the github issue

**Reply by nicitaacom** [reply]
Your answer was 2 years ago

I need to delete images from bucket every day

**Reply by GaryAustin1** [reply]
@nicitaacom This might give you some ideas on how to handle it https://github.com/GaryAustin1/supa-file-helper

**Reply by nicitaacom** [reply]
I pasted code from repository in SQL editor in spabase and got error - how to fix it?
```sql
ERROR:  2BP01: dependent privileges exist
HINT:  Use CASCADE to revoke them too.
CONTEXT:  SQL statement "grant usage on schema cron to postgres with grant option;
grant all on all functions in schema cron to postgres with grant option;
alter default privileges for user supabase_admin in schema cron grant all
    on sequences to postgres with grant option;
alter default privileges for user supabase_admin in schema cron grant all
    on tables to postgres with grant option;
alter default privileges for user supabase_admin in schema cron grant all
    on functions to postgres with grant option;
grant all privileges on all tables in schema cron to postgres with grant option;
revoke all on table cron.job from postgres;
grant select on table cron.job to postgres with grant option;
"
PL/pgSQL function inline_code_block line 3 at EXECUTE
SQL statement "do $$
begin
  execute
    replace(
      replace(
        replace(
          replace(
            pg_read_file(
'/etc/postgresql-custom/extension-custom-scripts/pg_cron/after-create.sql'            ),
            '@extname@', '''pg_cron'''          ),
          '@extschema@', '''extensions'''        ),
        '@extversion@', 'null'      ),     '@extcascade@', 'false'    );
exception
  when undefined_file then
    -- skip
end
$$"
```

**Reply by GaryAustin1** [reply]
This really is not the thread to debug something like that.
Seems like your cron has been set up and having some issue?  I've not seen an error like that whole thing you show.
You should try running the SQL a bit at a time, or disable and enable your cron extension in the UI to see if there are issues.

I ran the code down to setting up the function on a new instance and had no issue.
![image](https://github.com/supabase/supabase/assets/54564956/1b587bfe-0aa3-484c-b87d-7f971b1421ce)

**Reply by nicitaacom** [reply]
Thank you for advice
I disbled and enabled `pg_cron` and now I see 1 schedule

How to use it?
https://github.com/GaryAustin1/supa-file-helper/issues/2

**Reply by GaryAustin1** [reply]
As I said it is really an example of things you can do to call the storage API from SQL to delete files.  I show both pg_net and http calls to do bulk deletes.  And an example of a cron task to look for files that have been "marked" for deletion.  When I wrote the example it was before Supabase decoupled the files from having a foreign key link to auth.users, so I had to set that column to null in order to delete the user.  But also needed a way to know it was not an anon or super user owner which is also null.  There are probably better ways now.

But you said you need to delete files every day.  So you would use a cron task like I show and then figure out what files you want to delete (I assume you have an idea of which ones) and then use their path names to delete them similar to what I do.

This is not a guide as there are many different requirements but it shows the basics involved.

**Reply by GaryAustin1** [reply]
I'll also caution if you don't know SQL and what the code is up to by looking it at, this path may not be for you.  You need to be careful when dealing with Supabase schemas (like storage.objects) and you need to never just delete the rows in the table without calling the storage API or you will orphan file storage space forever.

**Comment by therealpurplemana** [comment]
Store your service role key and api url in the vault.  Then use this.  Makes it super easy to keep prod and dev separated.  Putting ngrok on top is even better.
```

CREATE OR REPLACE FUNCTION public.delete_from_public_url(public_url character varying)
 RETURNS character varying
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$
DECLARE
    supabase_url VARCHAR;
    service_role_key VARCHAR;
    bucket_name VARCHAR;
    pathname VARCHAR;
    headers_json JSONB;
    full_path TEXT;
BEGIN
    -- Retrieve the Supabase URL and Service Role Key from the vault
    SELECT decrypted_secret INTO supabase_url
    FROM vault.decrypted_secrets
    WHERE name = 'VAULT_SUPABASE_URL'
    LIMIT 1;

    SELECT decrypted_secret INTO service_role_key
    FROM vault.decrypted_secrets
    WHERE name = 'VAULT_SUPABASE_SERVICE_ROLE_KEY'
    LIMIT 1;

    -- Extract the bucket name and path from the public_url
    -- Assuming public_url format is "https://<supabase_url>/storage/v1/object/public/<bucket_name>/<path>"
    -- Remove the supabase URL part and split to get bucket and path
    full_path := substring(public_url from length(supabase_url || '/storage/v1/object/public/') + 1);
    bucket_name := split_part(full_path, '/', 1);
    pathname := substring(full_path from length(bucket_name) + 2);

    -- Prepare the authorization header
    headers_json := jsonb_build_object(
        'Authorization', 'Bearer ' || service_role_key
    );

    -- Perform the HTTP DELETE request and get the response
    PERFORM net.http_delete(
        url := supabase_url || '/storage/v1/object/' || bucket_name || '/' || pathname,
        headers := headers_json
    );

    -- Return success status, adjust as necessary based on actual response handling
    RETURN 'Deleted Successfully';  -- Placeholder, adjust based on your requirements
END;
$function$
;
```

EDIT, don't forget to secure it:

```
REVOKE ALL ON FUNCTION public.delete_from_public_url(character varying) FROM PUBLIC;
REVOKE ALL ON FUNCTION public.delete_from_public_url(character varying) FROM AUTHENTICATED;
REVOKE ALL ON FUNCTION public.delete_from_public_url(character varying) FROM ANON;
```


Or add a custom role for this

**Reply by GaryAustin1** [reply]
Although I don't doubt your function works, I caution anyone looking this post, unless I am missing something I can delete any file I know the name of in your system public bucket or not.  First there is no protection on the call so anon can call it.  2nd I can provide a public looking url with a  private bucket and path in your "public" format and it will delete it from the private bucket.

Storing in the vault is a good idea though.

**Reply by therealpurplemana** [reply]
Is that because it's in the public schema?  This is only invoked by trigger. 

REVOKE ALL ON FUNCTION public.delete_from_public_url(character varying) FROM PUBLIC;

I think this would ensure it can only run from service role

**Reply by GaryAustin1** [reply]
That is one issue, or you could certainly remove grants on the function from anon and authenticated (https://supabase.com/docs/guides/database/functions#function-privileges).    So yes with proper protection to service_role then you are safe.    The side note is that you pass in a "public" url but really all you need is bucket and path as the public part of that is just a placeholder and does not really mean public bucket, but that is more for other users understanding if they think it limits it to public buckets.

**Reply by GaryAustin1** [reply]
Revoking on PUBLIC is not enough.  You have to also revoke on authenticated and anon.

**Reply by therealpurplemana** [reply]
TY -- updated

**Comment by PaperPrototype** [comment]
just having the ability to delete folders would make this issue less of an issue. Right now I guess some people want to be able to delete a record with a ref to an image.

For example I have a course folder with images for each section of the course. I have to manually delete all the images for each section when I delete the course... which if you use folders you could delete all media inside of the course at once. 

https://github.com/supabase/storage/issues/207
